src/Servers/Kestrel/test/InMemory.FunctionalTests/Http2/Http2StreamTests.cs

```diff
@@ -1022,7 +1022,6 @@ public async Task ContentLength_Received_ReadViaPipes()
             {
                 var readResult = await context.Request.BodyPipe.ReadAsync();
                 Assert.Equal(12, readResult.Buffer.Length);
-                Assert.True(readResult.IsCompleted);
```

Comment:
This seems like a perf regression at the very least. Why isn't the ReadResult completed if the entire body has been read?

Reply:
Good point. I'll investigate more thoroughly.

Reply:
I just realized this was a new test. That makes it not as bad. It'd be nice if this wasn't racy though. It could certainly hurt request body consumption perf.

Reply:
Yeah. I didn't think it was racy either 🙃

Reply:
Looked into it. This code path is inherently racy, depending on when we call Complete on the Writer. In Http2Stream, we call complete in two places, OnRequestProcessingEnded and OnEndStreamReceived. These calls will happen either before, during or after the call to ReadAsync, depending on how quickly we call ReadAsync. 

As potentially an optimization for Http2, we can Complete the pipe when we received Content-Length amount of data. I don't believe I see us doing that today.